samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -218,6 +219,15 @@ class TaskInstance(
     trace("Checkpointing offsets for taskName: %s" format taskName)
 
     offsetManager.writeCheckpoint(taskName, checkpoint)
+
+    if (checkpoint != null) {
+      checkpoint.getOffsets.asScala
+        .filter { case (ssp, _) => intermediateStreams.contains(ssp.getStream) } // Only delete data of intermediate streams
```

Comment:
We can delete committed messages from intermediate stream because, in case the job fails, the job can be restarted again and regenerate all the data in the intermediate stream. However, if we delete committed messages from input stream, then if the job fails, the restarted job may not be able to reproduce the result because data in the input streams is deleted.

And yet, we can add this feature in low level API. We would need users to explicitly mark their input streams as "intermediate". I guess the goal of doing this only in high level API is that our long term goal is for most users to use the high level API. Maybe we can implement this in low level API in a separate patch if needed?

Reply:
As we discussed offline, let's decouple this feature from intermediate streams, since it can apply to all input streams. We can auto-write this config for intermediate streams from the Planner for high level API apps. @nickpan47 also agrees.

Reply:
@prateekm Sure. I have updated the patch that allows user to configure whether a stream's committed messages can be deleted.